### PR TITLE
Fallback to JPEG instead of PNG if image blob is opaque

### PR DESCRIPTION
**Pull request by brenogazzola**
### Summary
Active Storage representations currently fallback to PNG if the image blob is not in the `web_image_content_types`. While this ensures that the images are viewable in any browser, it generates rather large files that can eat through users data plans if they are viewing a large gallery, and make the page take much longer to load then necessary.

This PRs adds opaqueness detection to the image analyzers and uses this information to fallback to JPEG instead of PNG.

### Other Information
An image will be considered opaque is it has no alpha channel, or if all pixels have a specific value as their alpha (depends on the lib, but either `0` or `255`).

### CODE REVIEW

File: activestorage/lib/active_storage/analyzer/image_analyzer/vips.rb [hunk 1]
```diff
@@ -36,6 +36,11 @@ def rotated_image?(image)
         false
       end
 
+      def opaque?(image)
+        return true unless image.has_alpha?
+        image[image.bands - 1].min == 255
```
Comment (brenogazzola) [inline activestorage/lib/active_storage/analyzer/image_analyzer/vips.rb:41]: This from [an issue](https://github.com/libvips/libvips/issues/547) where jcupitt, maintainer of vips and ruby-vips, explained how to detect opaqueness.

File: activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb [hunk 2]
```diff
@@ -32,5 +32,12 @@ def read_image
       def rotated_image?(image)
         %w[ RightTop LeftBottom TopRight BottomLeft ].include?(image["%[orientation]"])
       end
+
+      def opaque?(image)
+        return true unless image.data["channelDepth"].key?("alpha")
+
+        value = /7.\d/.match?(image.data["version"]) ? 255 : 0
+        image.data["channelStatistics"]["alpha"]["mean"] == value
```
Comment (brenogazzola) [inline activestorage/lib/active_storage/analyzer/image_analyzer/image_magick.rb:40]: This is a lot of trial and error, and I have a problem here. For ImageMagick before 7 the value was `0`. For ImageMagick 7 and later the value is `255`. This PR works in both situations, but I don't know how to properly test for that in the CI.

**Comment by zzak** [comment]
Just a minor comment, but should this be a single comment + changelog?

**Comment by brenogazzola** [comment]
@zzak Changed. Should PRs always be a single CHANGELOG? I remember seeing one PR where someone said "squash into a few significant commits" and thought this would apply to changelogs too.

Sorry if this is obvious, I'm still trying to figure out the etiquette on contributing.

**Comment by zzak** [comment]
@brenogazzola Yes, sorry for the late reply. It's better to squash all commits (including the changelog) to a single commit. :bow:

**Comment by brenogazzola** [comment]
@zzak No worries. I’m here trying to pay back all the good the Rails community did for me, so I’m definitely not going to complain about this. 😜

If there’s anything else besides docs and possible upstreaming I can help with, just tell me 🤗

**Comment by brenogazzola** [comment]
Closed. CND's like Cloudflare will serve more modern file formats to browsers that support it, and with evergreen browsers, that's most of them.

**Pull request closed without merging**